### PR TITLE
[Don't merge] Stress test for CircularBuffer

### PR DESCRIPTION
**Pull request by reyang**
This is a stress test tool for the Circular Buffer [#1085].

```shell
dotnet run --configuration=Release
```

The test is writing to a circular buffer from N parallel tasks, where N is the number of logical processors.
Each writer is inserting monotonic numbers to the circular buffer, starting from `1` to `bound` inclusively (`[1, bound]`). If the buffer is full, the writer will retry until success.
A single task is reading from the circular buffer, calculating the total sum of the numbers.
Near real-time statistics are displayed in the console window title.

When all the writer/reader tasks finished, the `result` value should be `0`, otherwise it indicates a bug in the implementation.

Test buffer size can be changed to simulate different situation.

For example, changing the buffer size to a smaller number would simulate the insertion failure/retry case:

```csharp
var buffer = new CircularBuffer<Item>(10000);
```

image

### CODE REVIEW

**Comment by reyang** [comment]
Test buffer size can be changed to simulate different situation.

For example, changing the buffer size to a smaller number would simulate the insertion failure/retry case:

```csharp
var buffer = new CircularBuffer<Item>(10000);
```

image

**Comment by reyang** [comment]
Need help from folks to stress it on different architecture / platforms with different combinations of settings.

File: docs/trace/stress/Program.cs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    private static long cntEvents = 0;
+
+    public static int Main()
+    {
+        return EntryPoint();
+    }
+
+    private static int EntryPoint()
+    {
+        var cntWriter = Environment.ProcessorCount; // configure the number of writers
```
Comment (ThomsonTan) [inline docs/trace/stress/Program.cs:37]: Set `cntWriter` as `Environment.ProcessorCount - 1` and reserve one core for reader?

Comment (reyang) [reply docs/trace/stress/Program.cs:37]: That would make 0 writer on a single core machine 😆

Comment (ThomsonTan) [reply docs/trace/stress/Program.cs:37]: That's a good edge case, perhaps set 1 as minimum.😆

**Comment by ThomsonTan** [comment]
> Need help from folks to stress it on different architecture / platforms with different combinations of settings.

Ran the stress on x64 (12 cores 3.6GHz, Intel W-2133) and Surface Pro X (3.0GHz 8 cores ARM64) with .NET core 5.0 preview for ARM64, stress **passed** for more than 10 runs on each platform.

BTW: the race condition ratio of below line is ~1% (means failed to add 1 event among 100 events) on X64 and ~7% on ARM64.
>Interlocked.CompareExchange(ref this.head, headSnapshot + 1, headSnapshot) == headSnapshot

**Comment by reyang** [comment]
> Ran the stress on x64 (12 cores 3.6GHz, Intel W-2133) and Surface Pro X (3.0GHz 8 cores ARM64) with .NET core 5.0 preview for ARM64, stress **passed** for more than 10 runs on each platform.

Why would you have so many powerful 💪 machines? 🤣

**Pull request closed without merging**